src/logging.rs

```diff
@@ -0,0 +1,157 @@
+use std::fs::File;
+use std::io::Write;
+use std::os::unix::io::FromRawFd;
+use std::ffi::CString;
+use std::os::unix::net::UnixDatagram;
+use std::os::unix::io::{AsRawFd, IntoRawFd};
+use nix::sys::socket::{ControlMessage, MsgFlags, sendmsg};
+use nix::sys::memfd::MemFdCreateFlag;
+use nix::sys::memfd::memfd_create;
+use nix::fcntl::*;
+
+use crate::errors::*;
+
+/// Default path of the journald AF_UNIX datagram socket.
+const SD_JOURNAL_SOCK_PATH: &str = "/run/systemd/journal/socket";
+
+/// Log priority values.
+///
+/// See `man 3 syslog`.
+#[derive(Debug)]
+#[repr(u8)]
+pub enum Priority {
+    /// system is unusable
+    Emergency = 0,
+    /// action must be taken immediately
+    Alert,
+    /// critical conditions
+    Critical,
+    /// error conditions
+    Error,
+    /// warning conditions
+    Warning,
+    /// normal, but significant, condition
+    Notice,
+    /// informational message
+    Info,
+    /// debug-level message
+    Debug,
+}
+
+impl std::convert::From<Priority> for u8 {
+    fn from(p: Priority) -> Self {
+        match p {
+            Priority::Emergency => 0,
+            Priority::Alert => 1,
+            Priority::Critical => 2,
+            Priority::Error => 3,
+            Priority::Warning => 4,
+            Priority::Notice => 5,
+            Priority::Info => 6,
+            Priority::Debug => 7,
+        }
+    }
+}
+
+#[inline(always)]
+fn is_valid_char(c: char) -> bool {
+    c.is_uppercase() || c.is_numeric() || c == '_'
+}
+
+/// The variable name must be in uppercase and consist only of characters, 
+/// numbers and underscores, and may not begin with an underscore.
+fn is_valid_field(input: &str) -> bool {
+    if input.is_empty() {
+        return false;
+    }
+
+    if !input.chars().all(is_valid_char) {
+        return false;
+    }
+
+    if input.starts_with('_') {
+        return false;
+    }
+
+    true
+}
+
+fn add_field_and_payload(data: &mut String, field: &str, payload: &str) {
+    if is_valid_field(field) {
+        let field_payload = format!("{}={}\n", field, payload);
+        data.push_str(&field_payload);
+    }
+}
+
+/// Print a message to the systemd journal with the given priority.
+pub fn sd_journal_print(priority: Priority, msg: &str) -> Result<()> {
+    let sock = UnixDatagram::unbound()?;
+    sock.connect(SD_JOURNAL_SOCK_PATH)?;
```

Comment:
The C library has a different behavior here:
```
If systemd-journald(8) is not running (the socket is not present), those functions do nothing, and also return 0.
```

Perhaps we can check for the path presence, and change the return type into a `Result<bool>` to signal whether the message has been sent or not?

Reply:
Would it be better to return an `Err` if the socket path does not exist?

Reply:
It looks like the C library explicitly does not consider that as an error. I'd stick to that.

Reply:
Sorry, I misread the man page. Would `Result<usize>` be better?

Reply:
Mmmh, actually all of this logic becomes very race-y. Let's keep it as-is for the moment, I'll think about this a bit more for a followup.

Reply:
Sounds good. For future use: I was thinking about using lazy_static to test the presence of the socket path